### PR TITLE
Add tool to check OCP for missing @since tag

### PR DESCRIPTION
**Pull request by MorrisJobke**
- gets executed in autotest.sh and fails if some PHPDoc is missing
- doesn't check classes/methods that are marked as `@deprecated`
- iterates over all .php files in lib/public

Let's build more tools to force ourselves to a proper coding style \o/

cc @DeepDiver1975 @rullzer @PVince81 @LukasReschke @nickvergessen @schiesbn @karlitschek 

Currently it fails. I will push a second commit to this PR that fixes the current missing PHPDocs. I didn't pushed both together to test if the Jenkins setup works ;)

### CODE REVIEW

**Comment by MorrisJobke** [comment]
cc @Raydiation

**Comment by LukasReschke** [comment]
This is very cool.

**Comment by nickvergessen** [comment]
I think it's a fine solution

**Comment by MorrisJobke** [comment]
```
00:01:31.512 Using PHP executable /home/vagrant/.phpenv/shims/php
00:01:31.648 Parsing all files in lib/public for the presence of @since or @deprecated on each method...
00:01:31.648 
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\IUserManager::clearBackends
00:01:32.018 PHPDoc is needed for OCP\Files\LockNotAcquiredException::__construct
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\Files\LockNotAcquiredException::__toString
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\Files\Folder::newFile
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\Defaults::__construct
00:01:32.018 PHPDoc is needed for OCP\AppFramework\Db\Entity::setter
00:01:32.018 PHPDoc is needed for OCP\AppFramework\Db\Entity::getter
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\AppFramework\Db\Entity::markFieldUpdated
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\AppFramework\Db\Entity::addType
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\AppFramework\Db\DoesNotExistException::__construct
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\AppFramework\Db\MultipleObjectsReturnedException::__construct
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\AppFramework\Db\Mapper::isAssocArray
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\AppFramework\Db\Mapper::getPDOType
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\IUser::getHome
00:01:32.018 @since or @deprecated tag is needed in PHPDoc for OCP\IUser::canChangePassword
00:01:32.018 
00:01:32.310 Build step 'Execute shell' marked build as failure
```

Works :dancer:

File: autotest.sh [hunk 1]
```diff
@@ -89,6 +89,9 @@ if [ "$1" ]; then
 	fi
 fi
 
+# check for the presence of @since in all OCP methods
+$PHP tests/OCPSinceChecker.php
```
Comment (DeepDiver1975) [inline autotest.sh]: what about placing this script in the build folder - we have the license script in there as well

Comment (MorrisJobke) [inline autotest.sh]: ah ... didn't know about that. I will move it.

**Comment by MorrisJobke** [comment]
Because @nickvergessen asked: Is the execution of this script before the unit test execution okay? Or should we move it at the end? I added it there because then you see very quickly that something in OCP is undocumented and no time for phpunit test execution is wasted.

**Comment by DeepDiver1975** [comment]
> Because @nickvergessen asked: Is the execution of this script before the unit test execution okay? Or should we move it at the end? I added it there because then you see very quickly that something in OCP is undocumented and no time for phpunit test execution is wasted.

fine with me this way - THX

**Comment by DeepDiver1975** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
iqtce

**Comment by MorrisJobke** [comment]
jenkins #17043

**Comment by DeepDiver1975** [comment]
> jenkins #17043

thx

**Comment by MorrisJobke** [comment]
:rocket: Test PASSed.:rocket:
chuck

**Comment by MorrisJobke** [comment]
Reviewers? :)

**Comment by Raydiation** [comment]
:+1: Nothing worth mentioning and since this is a single file, there won't be any impact on other parts of the code.

File: lib/public/appframework/db/entity.php [hunk 2]
```diff
@@ -92,6 +92,10 @@ public function resetUpdatedFields(){
 		$this->_updatedFields = array();
 	}
 
+	/**
+	 * Generic setter for properties
+	 * @since 7.0.0
+	 */
 	protected function setter($name, $args) {
```
Comment (nickvergessen) [inline lib/public/appframework/db/entity.php]: missing parameters from the docs, fix incoming in new branch? :grinning:

File: lib/public/appframework/db/entity.php [hunk 3]
```diff
@@ -112,6 +116,10 @@ protected function setter($name, $args) {
 		}
 	}
 
+	/**
+	 * Generic getter for properties
+	 * @since 7.0.0
+	 */
 	protected function getter($name) {
```
Comment (nickvergessen) [inline lib/public/appframework/db/entity.php]: missing parameters from the docs, fix incoming in new branch? :grinning:

File: lib/public/files/locknotacquiredexception.php [hunk 4]
```diff
@@ -41,12 +41,20 @@ class LockNotAcquiredException extends \Exception {
 	/** @var integer $lockType The type of the lock that was attempted */
 	public $lockType;
 
+	/**
+	 * @since 7.0.0
+	 */
 	public function __construct($path, $lockType, $code = 0, \Exception $previous = null) {
```
Comment (nickvergessen) [inline lib/public/files/locknotacquiredexception.php]: missing parameters from the docs, fix incoming in new branch? :grinning:

**Comment by nickvergessen** [comment]
:+1: fine by me

**Comment by MorrisJobke** [comment]
@nickvergessen Feel free to fix the parameters in another PR ;)

**Pull request merged**